File: mtgSearch/src/main/java/com/dbottillo/mtgsearchfree/interactors/PlayerInteractorImpl.java
package com.dbottillo.mtgsearchfree.interactors;

import com.dbottillo.mtgsearchfree.model.Player;
import com.dbottillo.mtgsearchfree.model.storage.PlayersStorage;
import com.dbottillo.mtgsearchfree.util.LOG;
import com.dbottillo.mtgsearchfree.util.Logger;

import java.util.List;

import rx.Observable;

public class PlayerInteractorImpl implements PlayerInteractor {

    private final PlayersStorage storage;
    private final Logger logger;

    public PlayerInteractorImpl(PlayersStorage storage, Logger logger) {
        this.logger = logger;
        this.storage = storage;
        logger.d("created");
    }

    public Observable<List<Player>> load() {
        logger.d("loadSet");
        return Observable.just(storage.load());
    }

    @Override
    public Observable<List<Player>> addPlayer(Player player) {
        logger.d("add " + player);
        return Observable.just(storage.addPlayer(player));
    }

    @Override
    public Observable<List<Player>> editPlayer(Player player) {
        logger.d("edit " + player);
        return Observable.just(storage.editPlayer(player));
    }

    @Override
    public Observable<List<Player>> editPlayers(List<Player> players) {
        logger.d("update players " + players.toString());
        return Observable.just(storage.editPlayers(players));
    }

    @Override
    public Observable<List<Player>> removePlayer(Player player) {
        logger.d("remove " + player);
        return Observable.just(storage.removePlayer(player));
    }

}

File: MTGSearch/src/main/kotlin/com/dbottillo/mtgsearchfree/interactors/PlayerInteractorImpl.kt
package com.dbottillo.mtgsearchfree.interactors

import com.dbottillo.mtgsearchfree.interactor.SchedulerProvider
import com.dbottillo.mtgsearchfree.model.Player
import com.dbottillo.mtgsearchfree.storage.PlayersStorage
import com.dbottillo.mtgsearchfree.util.Logger
import io.reactivex.Observable

class PlayerInteractorImpl(
    private val storage: PlayersStorage,
    private val schedulerProvider: SchedulerProvider,
    private val logger: Logger
) : PlayerInteractor {

    init {
        logger.d("created")
    }

    override fun load(): Observable<List<Player>> {
        logger.d("loadSet")
        return Observable.fromCallable { storage.load() }
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
    }

    override fun addPlayer(): Observable<List<Player>> {
        logger.d("add player")
        return Observable.fromCallable { storage.addPlayer() }
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
    }

    override fun editPlayer(player: Player): Observable<List<Player>> {
        logger.d("edit " + player)
        return Observable.fromCallable { storage.editPlayer(player) }
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
    }

    override fun editPlayers(players: List<Player>): Observable<List<Player>> {
        logger.d("update players " + players.toString())
        return Observable.fromCallable { storage.editPlayers(players) }
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
    }

    override fun removePlayer(player: Player): Observable<List<Player>> {
        logger.d("remove " + player)
        return Observable.fromCallable { storage.removePlayer(player) }
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
    }
}
